### ISSUE TITLE
Can you enable travis and codecov in the backend?

### ISSUE THREAD

**Issue by cdancy** [bmuschko/gradle-kubernetes-plugin#1]
Finally am getting around to pickup the work here to start developing this plugin. @bmuschko can you log into travis and codecov and ensure we can build successfully on those services? Can you also copy over the same creds/env-vars from `gradle-docker-plugin` so that we can mimic the release process? Also might not be a bad idea to create a `gradle-kubernetes-plugin` repo on Bintray as well so that we can publish something to.

@orzeh if you want to start helping out in any way you can easily stand-up a kubernetes cluster on your local machine using docker containers. On mac this comes out of the box if you're using their `edge` releases. If not, or on another machine, there are loads of tutorials on how to use minikube or localkube which amounts to the same thing.

**Comment by bmuschko** [comment]
@cdancy done

**Comment by cdancy** [comment]
@bmuschko Bintray as well?

**Comment by bmuschko** [comment]
@cdancy I believe the package will be created when you publish for the first time. Let me know if that should not work the way I think it works.

**Comment by cdancy** [comment]
@bmuschko Ok I'll tackle that when I come to it then. I vaguely remember needing to import a project from github but maybe that's not the case.

**Comment by cdancy** [comment]
Closing as ISSUE as now resolved.

**Comment by cdancy** [comment]
@orzeh @bmuschko if you can spare a bit of time I've got the general project structure in place and would love if you can give things a look over and see if they make sense or if you have any general questions/comments/concerns.

**Comment by bmuschko** [comment]
@cdancy Thanks, I will try to look over it by the end of the week. How do you prefer to receive feedback?

**Comment by cdancy** [comment]
@bmuschko open a new ISSUE or something else? Let me state a few design decisions here before folks go poking around as I've not fully documented my intentions or some of the features yet.

The plugin itself works in largely the same way as the `gradle-docker-plugin`. I simplified things a bit and factored out lots of code to `traits` (more on that below). We're using the fabric8.io `kubernetes-client` library for interacting with kubernetes. They provide a shaded version as well so I opted to use that if only to get around many of the issues we have with the `gradle-docker-plugin`. Furthermore I used the `Jcl` (java class loader) library to create a separate/isolated class-loader for the kubernetes-client classpath to further ensure that it won't clobber anything else put on the `buildscript` classpath.

One design goal that I was going with is to be able to "configure" each tasks properties/variables/etc instead of us having to code and expose these properties every time a user wants additional options. This is done through the `ConfigAware` trait, which gets applied to all tasks, and works something like this:
```
task someKubeTask (type: AwesomeKubernetesTask) {
    config {
        setPropertyOne('abcd')
        withSomeOtherProperty('efgh')
    }
}
```
The `kubernetes-client` instance has many ... domain?!? ... methods which return an Object of said domain, which can be further configured in some way, and then finally executed for a result. The `config` closure above gets applied on each domain object and then we execute the given command. Thought this was a good way of empowering the user to configure things however they want and getting us out of their way. Also seemed to be the only sane route less we want to expose a thousand properties that the client provides which could become a nightmare.

Another design idea is the `ResponseAware` trait which also gets applied to all tasks. As each execution in the `kubernetes-client` returns "something" I wanted a way for all tasks to return whatever that something is so that downstream tasks have a standard way of accessing and working with it in a gradle context. An example might look something like:
```
task getMyPods(type: ListPods) {
    // yes I brought along the reactive-stream methods as well as I find them to be very useful.
    onNext {
        logger.quiet "Found pod: ${it}"
    }
}

task downstreamTask(dependsOn: getMyPods) {
    doLast {
        def podsResponse = getMyPods.response()
        // do some additional work with podsResponse object
    }
}
```
As the `response()` could literally return anything, and thus potentially confusing the user, I was going to add an additional column to each tasks table which states not only what the `response()` object is but also what domain object the `config {}` closure further above is actually configuring and give direct links to the source.

Those are the few big ideas I'm going with everything else should be pretty self-explanatory. Travis-CI is already wired up and I've written a few basic functionalTests to get things working. The source itself contains loads of documentation and design-ideas itself that should further explain anything here I may have missed.

But yeah as far as feedback we could either do things here, in a follow-up ISSUE, or maybe setup a gitter chat? Anything works for me just wanted to get some input on the initial setup and if things make sense and if there was anything that could be done better or if I just completely missed on something :)
